zio-kafka/src/main/scala/zio/kafka/consumer/internal/Runloop.scala

```diff
@@ -57,113 +62,152 @@ private[consumer] final class Runloop private (
+          // When the queue is empty we still need to call commit (with 0 offsets) so that earlier commits can complete.
+          // We can not use ZStream.fromQueue because that will emit nothing when the queue is empty.
+          // TODO: add timeout; we can't postpone the rebalance for ever
+          ZStream
+            .fromZIO(commitQueue.takeAll)
+            .repeat(Schedule.forever)
+            .tap(doCommitsFromRebalanceListener(rebalanceConsumer))
+            .takeUntilZIO(_ => ZIO.forall(streamsToEnd)(_.isCompleted))
+            .runCollect
+            .map(_.flatten)
+        } else {
+          ZIO.succeed(Chunk.empty)
+        }
+    } yield pendingCommits
+
+  /**
+   * Handle commits while waiting for revoked streams to end.
+   *
+   * We need to ensure the revoked streams can end. This is only possible if the commits these streams started complete.
+   * The commits complete when the callback is invoked. The callback is invoked when the underlying consumer polls the
+   * broker. This can be achieved by invoking `commitAsync`. Even when we pass no offsets, the broker will be polled and
+   * callbacks will be called.
+   */
+  private def doCommitsFromRebalanceListener(
+    rebalanceConsumer: RebalanceConsumer
+  )(commits: Chunk[Commit]): Task[Unit] = {
+    val (offsets, callback, onFailure) = asyncCommitParameters(commits)
+    ZIO.logTrace(s"Async commit of ${offsets.size} offsets for ${commits.size} commits") *>
+      rebalanceConsumer
+        .commitAsync(offsets, callback)
+        .catchAll(onFailure)
+        .unit
+  }
 
-  private def doCommit(cmd: Commit): UIO[Unit] = {
-    val offsets   = cmd.offsets.map { case (tp, offset) => tp -> new OffsetAndMetadata(offset + 1) }
-    val cont      = (e: Exit[Throwable, Unit]) => cmd.cont.done(e).asInstanceOf[UIO[Unit]]
-    val onSuccess = cont(Exit.unit) <* diagnostics.emitIfEnabled(DiagnosticEvent.Commit.Success(offsets))
+  private def handleCommits(state: State, commits: Chunk[Commit]): Task[State] = {
+    val (offsets, callback, onFailure) = asyncCommitParameters(commits)
+    consumer.withConsumerZIO { c =>
+      // We don't wait for the completion of the commit here, because it
+      // will only complete once we poll again.
+      ZIO.attempt(c.commitAsync(offsets.asJava, callback))
+    }
+      .catchAll(onFailure)
+      .unit
+      .as(state.addCommits(commits))
+  }
+
+  private def asyncCommitParameters(
+    commits: Chunk[Commit]
+  ): (Map[TopicPartition, OffsetAndMetadata], OffsetCommitCallback, Throwable => UIO[Unit]) = {
+    val offsets = commits.foldLeft(Map.empty[TopicPartition, Long]) { case (acc, commit) =>
+      commit.offsets.foldLeft(acc) { case (acc, (tp, offset)) =>
+        acc.updated(tp, acc.get(tp).map(_ max offset).getOrElse(offset))
+      }
+    }
+    val offsetsWithMetaData = offsets.map { case (tp, offset) => tp -> new OffsetAndMetadata(offset + 1) }
```

Comment:
This value is only useful when the diagnostic is enabled, so most probably useless in prod for our users
```suggestion
    def offsetsWithMetaData = offsets.map { case (tp, offset) => tp -> new OffsetAndMetadata(offset + 1) }
```